dataplane/snappi/sai_dataplane.py

```diff
@@ -0,0 +1,334 @@
+        # Configure a new API instance where the location points to controller
+        # Ixia-C:       location = "https://<tgen-ip>:<port>"
+        # IxNetwork:    location = "https://<tgen-ip>:<port>", ext="ixnetwork"
+        # TRex:         location =         "<tgen-ip>:<port>", ext="trex"
+        if self.mode == 'ixnet':
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False, ext='ixnetwork')
+        elif self.mode == 'trex':
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False, ext='trex')
+        else:
+            self.api = snappi.api(location=self.exec_params['controller'], verify=False)
+        logging.info("%s Starting connection to controller... " % time.strftime("%s"))
+
+        # Create an empty configuration to be pushed to controller
+        self.configuration = self.api.config()
+
+        # Configure two ports where the location points to the port location:
+        # Ixia-C:       port.location = "localhost:5555"
+        # IxNetwork:    port.location = "<chassisip>;card;port"
+        # TRex:         port.location = "localhost:5555"
+        for pid, port in enumerate(self.exec_params['port_groups']):
+            location = port.get('location', f"localhost:{BASE_TENGINE_PORT+pid}")
+            self.configuration.ports.port(name=port[port['init']], location=location)
+
+        cap = self.configuration.captures.capture(name="c1")[-1]
+        cap.port_names = [p.name for p in self.configuration.ports]
+        cap.format = cap.PCAP
+
+        self.dataplane = self
+
+    def remove(self):
+        pass
+
+    def setUp(self):
+        super().setUp()
+        self.set_config()
+        self.start_capture()
+
+    def tearDown(self):
+        super().tearDown()
+        self.stop_capture()
+        self.configuration.flows.clear()
+        self.flows.clear()
+
+    @staticmethod
+    def api_results_ok(results):
+        if hasattr(results, 'warnings'):
+            return True
+        else:
+            print(f'Results={results}')
+            return False
+
+    @staticmethod
+    def check_warnings(response):
+        if response.warnings:
+            print("Warning: %s" % str(response.warnings))
+
+    @staticmethod
+    def get_capture_port_names(cfg):
+        """
+        Returns name of ports for which capture is enabled.
+        """
+        names = []
+        for cap in cfg.captures:
+            if cap._properties.get("port_names"):
+                for name in cap.port_names:
+                    if name not in names:
+                        names.append(name)
+        return names
+
+    def set_config(self, return_time=False):
+        start_time = time.time()
+        logging.info('Setting config ...')
+        res = self.api.set_config(self.configuration)
+        assert self.api_results_ok(res), res
+        if len(res.warnings) > 0:
+            logging.info('Warnings in set_config : {}'.format(res.warnings))
```

Comment:
Couple of notes:
- Upon error, exception is raised. If this is not done across different implementation, then it's a bug in the implementation
- timers should be guarded by `try/finally` to guarantee execution
- Warnings could technically be logged inside a different func

Here's how we do it: https://github.com/open-traffic-generator/conformance/blob/dd628aa14c2ae5e4f4d8d4858b07a6acc50ce6d0/helpers/otg/otg.py#L113

Reply:
fixed